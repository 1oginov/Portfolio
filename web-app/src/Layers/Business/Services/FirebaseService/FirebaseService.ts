import firebase from 'firebase/app';
import 'firebase/auth';

// TODO: Move to Infrastructure layer.
import * as C from 'Constants';
import { VibrantPalette } from 'Layers/Business/Services/FirebaseFunctionsService/VibrantPalette';

export const getAuthProvider = (provider: string): firebase.auth.AuthProvider => {
  switch (provider) {
    case 'google':
      return new firebase.auth.GoogleAuthProvider();

    default:
      throw new Error('Unknown auth provider');
  }
};

export const getPositionsCollectionPath = (userId: string): string => (
  `${C.FIRESTORE_USERS_COLLECTION}/${userId}/${C.FIRESTORE_POSITIONS_COLLECTION}`
);

export const getPositionsCollection = (
  firebaseInstance: firebase.app.App,
  userId: string,
): firebase.firestore.CollectionReference => (
  firebaseInstance.firestore()
    .collection(C.FIRESTORE_USERS_COLLECTION)
    .doc(userId)
    .collection(C.FIRESTORE_POSITIONS_COLLECTION)
);

export const getPositionDocument = (
  firebaseInstance: firebase.app.App,
  userId: string,
  positionId: string,
): firebase.firestore.DocumentReference => getPositionsCollection(firebaseInstance, userId).doc(positionId);

export const getVibrantColor = (
  vibrantPalette: VibrantPalette | null,
  type: 'vibrant' | 'light' | 'dark' | 'muted' | 'lightMuted' | 'darkMuted' = 'vibrant',
  opacity = 1,
): string => {
  if (vibrantPalette === null) {
    return 'none';
  }

  let color;

  switch (type) {
    case 'light':
      color = vibrantPalette.LightVibrant;
      break;
    case 'dark':
      color = vibrantPalette.DarkVibrant;
      break;
    case 'muted':
      color = vibrantPalette.Muted;
      break;
    case 'lightMuted':
      color = vibrantPalette.LightMuted;
      break;
    case 'darkMuted':
      color = vibrantPalette.DarkMuted;
      break;
    default:
      color = vibrantPalette.Vibrant;
      break;
  }

  return `rgba(${color.rgb[0]}, ${color.rgb[1]}, ${color.rgb[2]}, ${opacity})`;
};
