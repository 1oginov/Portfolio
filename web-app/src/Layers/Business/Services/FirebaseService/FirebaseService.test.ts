import { getAuthProvider, getPositionsCollectionPath } from './FirebaseService';

describe('getAuthProvider', () => {
  it('returns Google auth provider if "google" passed', () => {
    expect(getAuthProvider('google')).toHaveProperty('providerId', 'google.com');
  });

  it('throws error matching snapshot if unknown provider passed', () => {
    expect(() => getAuthProvider('unknown')).toThrowErrorMatchingSnapshot();
  });
});

describe('getPositionsCollectionPath', () => {
  it('matches snapshot', () => {
    expect(getPositionsCollectionPath('userId')).toMatchSnapshot();
  });
});
