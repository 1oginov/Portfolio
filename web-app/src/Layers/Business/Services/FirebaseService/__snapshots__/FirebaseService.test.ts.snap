// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAuthProvider throws error matching snapshot if unknown provider passed 1`] = `"Unknown auth provider"`;

exports[`getPositionsCollectionPath matches snapshot 1`] = `"users/userId/positions"`;
